gossipsub.go

```diff
@@ -239,6 +267,90 @@ func (gs *GossipSubRouter) handlePrune(p peer.ID, ctl *pb.ControlMessage) {
 			gs.tracer.Prune(p, topic)
 			delete(peers, p)
 			gs.untagPeer(p, topic)
+			gs.addBackoff(p, topic)
+			px := prune.GetPeers()
+			if len(px) > 0 {
+				gs.pxConnect(px)
```

Comment:
I have two concerns:

1. What if a single peer sends a _bunch_ of prunes? Could they cause us to launch a ton of goroutines (and crash)?
2. Can't a peer use this to trick us into connecting to a _ton_ of (potentially non-useful) peers? We should only try connecting to the number of peers we actually _need_ (one?).

Ideally, we'd stick these peers in a set of known peers for the topic, connecting to them as-needed whenever we're pruned or we disconnect from a peer.

Reply:
1. We must be already grafted, we ignore prunes that don't correspond to a peer on our mesh.
    So a peer can't make us launch an arbitrary number of goroutines by sending us prunes;
    at most he can make us launch a single goroutine for each topic we belong to its mesh.
    Not much of a vector I think.

Reply:
2. We could have a (legitimate) prune listing an arbitrary number of useless peers; we could limit the number of peers we connect to.

Reply:
For 2 I added a check that limits the number of connections to at most `GossipSubPrunePeers`,
so this should address the concern. We do want more than 1 peer in general, to expand our potential peers as much as possible.

Reply:
Got it. SGTM.

Reply:
Ok, I slept on it and there _is_ a DoS vector: A malicious peer could send us in sequence GRAFT/PRUNE and cause us to spawn a goroutine; it could simply be sitting there sending GRAFT/PRUNE ad infinum, causing us to spawn a goroutine for each pair.
Granted, the effect is limited in how many goroutines it can fit inside the 30s window for the connection timeout, but it's still nasty.
I will rewrite the connection logic to use a limited set of pending connections and goroutine-free scheduling.

Reply:
Implemented a connection scheduler, which limits the number of max pending connections too.